runtime/vm/threadhelp.cpp

```diff
@@ -78,140 +76,151 @@ validateTimeouts(J9VMThread* vmThread, I_64 millis, I_32 nanos)
-
+		J9VMTHREAD_SET_BLOCKINGENTEROBJECT(vmThread, vmThread, object);
+		object = NULL;
+		internalReleaseVMAccessSetStatus(vmThread, thrstate);
 #if defined(J9VM_OPT_CRIU_SUPPORT)
-	I_64 beforeWait = 0;
-	if (waitTimed) {
-		PORT_ACCESS_FROM_JAVAVM(javaVM);
-		beforeWait = j9time_nano_time();
-	}
+		/* Monitor wait time adjustment only supports CRIURestoreNonPortableMode which is default mode. */
+		BOOLEAN adjustmentRequired = waitTimed && (0 == javaVM->checkpointState.checkpointRestoreTimeDelta);
+		BOOLEAN singleThreadModeContinueWaitRequired = waitTimed && interruptable;
+		I_64 beforeWait = 0;
+		if (adjustmentRequired || singleThreadModeContinueWaitRequired) {
+			PORT_ACCESS_FROM_JAVAVM(javaVM);
+			beforeWait = j9time_nano_time();
+		}
 continueWait:
 #endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
-
-	if (interruptable) {
-		rc = omrthread_monitor_wait_interruptable(monitor, millis, nanos);
-	} else {
-		rc = omrthread_monitor_wait_timed(monitor, millis, nanos);
-	}
-
+		if (FALSE != interruptable) {
+			rc = omrthread_monitor_wait_interruptable(monitor, millis, nanos);
+		} else {
+			rc = omrthread_monitor_wait_timed(monitor, millis, nanos);
+		}
 #if defined(J9VM_OPT_CRIU_SUPPORT)
-	if (((J9THREAD_INTERRUPTED == rc)
-		|| (J9THREAD_PRIORITY_INTERRUPTED == rc))
-		&& J9_IS_SINGLE_THREAD_MODE(javaVM)
-	) {
-		if (waitTimed) {
-			PORT_ACCESS_FROM_JAVAVM(javaVM);
-			const I_32 oneMillion = 1000000;
-			I_64 waitedTimeNanos = j9time_nano_time() - beforeWait;
-			I_64 waitedTimeMillis = waitedTimeNanos / oneMillion;
-			if (waitedTimeMillis < millis) {
-				millis -= waitedTimeMillis;
-				waitedTimeNanos = (waitedTimeNanos % oneMillion);
-				if (waitedTimeNanos >= nanos) {
-					nanos = (I_32)(waitedTimeNanos - nanos);
+		BOOLEAN interruptedAtSingleThreadMode = ((J9THREAD_INTERRUPTED == rc) || (J9THREAD_PRIORITY_INTERRUPTED == rc)) && J9_IS_SINGLE_THREAD_MODE(javaVM);
+		if (interruptedAtSingleThreadMode
+			|| (adjustmentRequired && (0 != javaVM->checkpointState.checkpointRestoreTimeDelta))
+		) {
+			/* Two scenarios: in CRIU single thread mode, the thread is interrupted;
+			 * or a checkpoint has been taken, waited time adjustment might be required.
+			 */
+			BOOLEAN continueWaitRequired = TRUE;
+			if (waitTimed) {
+				PORT_ACCESS_FROM_JAVAVM(javaVM);
+				const I_32 oneMillion = 1000000;
+				I_64 waitedTimeNanos = j9time_nano_time() - beforeWait;
+				I_64 waitedTimeMillis = waitedTimeNanos / oneMillion;
+				if (waitedTimeMillis < millis) {
+					millis -= waitedTimeMillis;
+					waitedTimeNanos = waitedTimeNanos % oneMillion;
+					if (nanos >= waitedTimeNanos) {
+						nanos = (I_32)(nanos - waitedTimeNanos);
+					} else {
+						/* nanos/waitedTimeNanos are within [0, 999999] */
+						nanos = (I_32)(oneMillion - waitedTimeNanos + nanos);
+						millis -= 1;
+					}
+				} else if (interruptedAtSingleThreadMode) {
+					/* timed out, waiting another 10ms in single thread mode */
```

Comment:
if its a spurious interrupt it wouldn't be a time out. We should still check to see if there is time remaining and wait a least that amount

Reply:
There is a check `if (waitedTimeMillis < millis)` before this, the requested time `millis` is already finished, there might be still `nanos`, however if currently VM is in single threaded mode, `continueWait` is required, I picked `10ms` for next wait period.
If VM is not in single threaded mode, further checking for `nanos` is at L159 block and is to wait accordingly.